kimchi/src/circuits/expr.rs

```diff
@@ -520,56 +545,68 @@ impl<F: FftField> PolishToken<F> {
-                ForeignFieldModulus(i) => {
-                    if let Some(modulus) = c.foreign_field_modulus.clone() {
-                        stack.push(ForeignElement::<F, 3>::from_biguint(modulus.clone())[*i])
+            if skip_count > 0 {
+                skip_count -= 1;
+            } else {
+                use PolishToken::*;
+                match t {
+                    Alpha => stack.push(c.alpha),
+                    Beta => stack.push(c.beta),
+                    Gamma => stack.push(c.gamma),
+                    JointCombiner => {
+                        stack.push(c.joint_combiner.expect("no joint lookup was expected"))
+                    }
+                    EndoCoefficient => stack.push(c.endo_coefficient),
+                    Mds { row, col } => stack.push(c.mds[*row][*col]),
+                    ForeignFieldModulus(i) => {
+                        if let Some(modulus) = c.foreign_field_modulus.clone() {
+                            stack.push(ForeignElement::<F, 3>::from_biguint(modulus.clone())[*i])
+                        }
+                    }
+                    VanishesOnLast4Rows => stack.push(eval_vanishes_on_last_4_rows(d, pt)),
+                    UnnormalizedLagrangeBasis(i) => {
+                        stack.push(unnormalized_lagrange_basis(&d, *i, &pt))
+                    }
+                    Literal(x) => stack.push(*x),
+                    Dup => stack.push(stack[stack.len() - 1]),
+                    Cell(v) => match v.evaluate(evals) {
+                        Ok(x) => stack.push(x),
+                        Err(e) => return Err(e),
+                    },
+                    Pow(n) => {
+                        let i = stack.len() - 1;
+                        stack[i] = stack[i].pow(&[*n as u64]);
+                    }
+                    Add => {
+                        let y = stack.pop().ok_or(ExprError::EmptyStack)?;
+                        let x = stack.pop().ok_or(ExprError::EmptyStack)?;
+                        stack.push(x + y);
+                    }
+                    Mul => {
+                        let y = stack.pop().ok_or(ExprError::EmptyStack)?;
+                        let x = stack.pop().ok_or(ExprError::EmptyStack)?;
+                        stack.push(x * y);
+                    }
+                    Sub => {
+                        let y = stack.pop().ok_or(ExprError::EmptyStack)?;
+                        let x = stack.pop().ok_or(ExprError::EmptyStack)?;
+                        stack.push(x - y);
+                    }
+                    Store => {
+                        let x = stack[stack.len() - 1];
+                        cache.push(x);
+                    }
+                    Load(i) => stack.push(cache[*i]),
+                    SkipIf(feature, count) => {
+                        if !feature.is_enabled() {
+                            skip_count = *count;
+                            stack.push(F::zero());
```

Comment:
alternatively up there you can write:

```rust
while let Some(t) = toks.next() {
  // ...
    Skipif(feature, count) if !feature.is_enabled() => {
      toks.skip(count);
```

not sure I understand the `stack.push(F::zero())` yet though (still reading)

Reply:
interesting to see that this wouldn't work if you have a feature you're multiplying with something else (as pushing zero on the stack would cancel the feature). I guess we're not expecting this to happen but maybe good to keep in mind?